Cassandra improvements - tickets: 5506 5507 5508 - alignment of features

razvan-becheriu opened the pull request.

razvan-becheriu commented:
All unit tests Pass!
You can merge this PR!

The content of this PR includes:
implementation for tickets:   
5506: Cassandra host data source does not support host deletion
5507: Cassandra host backend doesn't store next-server, server-hostname and bootfile
5508: Cassandra host reservations don't support user context

implemented missing functionality in pgsql making PR9 unnecessary and resolving tickets:
4530: Add use of hardware address, type, and source to PostgreSQL lease manager

also added missing data from cql tables:
lease_hwaddr_source
0 |                   HWADDR_SOURCE_UNKNOWN
128 |              HWADDR_SOURCE_DOCSIS_MODEM

added missing tables in cql:
dhcp_option_scope
 scope_id |  scope_name  
----------+--------------
        0 | global
        1 | subnet
        2 | client-class
        3 | host

host_identifier_type
 type |    name    
------+------------
    0 | hw-address
    1 | duid
    2 | circuit-id
    3 | client-id
    4 | flex-id

added missing functionality in cql lease manager:
Lease4Collection getLeases4(SubnetID subnet_id) const;
Lease4Collection getLeases4() const;

added/enabled tests:
[ RUN      ] CQLHostMgrTest.getAll
[ RUN      ] CQLHostMgrTest.getAll4
[ RUN      ] CQLHostMgrTest.get4
[ RUN      ] CQLHostMgrTest.get6
[ RUN      ] CQLHostMgrTest.get6ByPrefix
[ RUN      ] CqlHostDataSourceTest.get4ByClientId
[ RUN      ] CqlHostDataSourceTest.usercontext
[ RUN      ] CqlHostDataSourceTest.multipleClientClasses4
[ RUN      ] CqlHostDataSourceTest.multipleClientClasses6
[ RUN      ] CqlHostDataSourceTest.multipleClientClassesBoth
[ RUN      ] CqlHostDataSourceTest.messageFields
[ RUN      ] CqlHostDataSourceTest.testMultipleHosts6
[ RUN      ] CqlLeaseMgrTest.getLeases4SubnetId
[ RUN      ] CqlLeaseMgrTest.getLeases4
[ RUN      ] CqlHostDataSourceTest.optionsReservations4
[ RUN      ] CqlHostDataSourceTest.optionsReservations6
[ RUN      ] CqlHostDataSourceTest.formattedOptionsReservations4
[ RUN      ] CqlHostDataSourceTest.formattedOptionsReservations6
[ RUN      ] CqlHostDataSourceTest.testMultipleHostsNoAddress4
[ RUN      ] CqlHostDataSourceTest.deleteByAddr4
[ RUN      ] CqlHostDataSourceTest.deleteById4
[ RUN      ] CqlHostDataSourceTest.deleteById4Options
[ RUN      ] CqlHostDataSourceTest.deleteById6
[ RUN      ] CqlHostDataSourceTest.deleteById6Options
[ RUN      ] PgSqlLeaseMgrTest.testRecreateLease4
[ RUN      ] PgSqlLeaseMgrTest.testRecreateLease6
[ RUN      ] PgSqlLeaseMgrTest.deleteExpiredReclaimedLeases6
[ RUN      ] PgSqlLeaseMgrTest.testLease6Mac
[ RUN      ] PgSqlLeaseMgrTest.testLease6HWTypeAndSource

Only one test remains disabled (for Cassandra, but it is enabled for the other back-ends) as this involves actually modifying the cassandra server configuration (TBD):
[  DISABLED     ] CqlHostDataSourceTest.testReadOnlyDatabase

Other changes incluse:
fix inconsistent spacing
fix spelling mistakes
fix typos
reordered function declarations and definitions to be consistent between back-ends/headers/source files

fxdupont commented:
As you know I replaced MD5 by FNV-1a as the hash function in the host backend code. I didn't include an obvious improvement because it could have impact outside the hash code itself (e.g. requires a schema update): hash returns a unsigned 64 bit integer which is cast into a signed 64 integer without any need and dropping the first bit. I open the Kea 5550 ticket about this but it was decided this should be included in this Pull Request.
Please answer to this comment so I can close the 5550 ticket.

razvan-becheriu commented:
Hi,

Some clarifications needed:
What do you mean by:
dropping the first bit.

and in the ticket comment:
Both because hashes are naturally unsigned, second because the cast drops the first bit.

First of all, the cast does not drop the sign bit, the result will be a negative number if the unsigned value is greater than the max signed integer value.

Second of all, I don't understand what answer do you need, because I don't understand the question.

Thanks,
Razvan

fxdupont commented:
I become paranoid with integer conversion in other languages: C++ does not drop the sign bit.
Nevertheless IMHO the hash return type should be cass_uint64_t, not cast_int64_t.

razvan-becheriu commented:
Hi,

No, we need to convert to cast_int64_t because the BindData function
receives data in an AnyArray, an array or boost::any and it calls the type
cast operator of cass_int32_t:

static CassError
CqlBindInt64(const boost::any& value,
             const size_t& index,
             CassStatement* statement) {
    return cass_statement_bind_int64(statement, index,

 *boost::any_cast<cass_int64_t*>(value));
}

If you add an cass_uint64_t boost::any, this will throw.

Don't worry, I have reviewed the impact and everything seems ok.
You can close the ticket.

Regards,
Razvan

On Wed, Feb 28, 2018 at 5:20 PM, fxdupont <notifications@github.com> wrote:

> I become paranoid with integer conversion in other languages: C++ does not
> drop the sign bit.
> Nevertheless IMHO the hash return type should be cass_uint64_t, not
> cast_int64_t.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/isc-projects/kea/pull/70#issuecomment-369271371>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/ASkW2ww6nE7mQgionoqLPcBewPecCoKIks5tZW3egaJpZM4SECZP>
> .
>



-- 
razvan becheriu
[image: qualitance.com]
senior software engineer
+40 (746) 090 415 <+40+(746)+090+415>, qualitance.com
<https://www.qualitance.com/>

fxdupont commented:
No problem. I closed 5550.

tomaszmrugalski commented:
I have merged the code to our internal master. It should appear on github in an hour or so.
First time did review comments on github and it seems I now can't close this PR...

I will close it as soon as I figure out what's preventing me from closing it.
Thanks for the patch and for covering other things.

razvan-becheriu commented:
To answer your questions (I am not sure you are going to like it).

The thing with the line wrap before the 'status' column is because in our development branch we have implemented other columns, like in the case of the draft-ietf-dhc-dhcp4o6-saddr-opt, we have added the 'sw_4o6_src_address' field in the database.
It would be nice if all column would be on a separate line (see CQL implementations), because, if in the future there is any need to add a new column, the differences will be easier to spot. I have aligned only the last column, the 'status' and I insert every new column in front of it. Something similar should be done in the every 'class' constructors, for the same reason.

Why have I added functionality in the PgSql in this Cassandra PR.
Well, I don't have a list of all features, so I use a diff tool (beyond compare) to compare between files (mysql_lease_mgr, pgsql_lease_mgr, cql_lease_mgr).
I have looked over the differences in the files for a couple of days and I made a list with all the features that were missing in Cassandra. Doing so, I came across the missing functionality in the PgSql implementation, so I've implemented it. It was too much of a trouble to create a new branch and switch between branches, but I am sorry if this made things harder for you to do the review. This is why I have mentioned to do the review per file, because mush of the differences were not related to other files.
I did the same for the unit tests. The whole bunch of spacing, comments, and function ordering is related to the fact that I have aligned all implementations. Now, if you do a diff between the different implementations, you will see that only the names (PgSql/MySql/Cql) are different. The implementation is the same. I also did this for the ipv4/ipv6 implementations (where it made sense - where the implementation should be the same).

Regarding LeaseStatsQueryPtr, well, the previous motivation included this change.

Regarding the swap between the ipv4 and ipv6, in all the implementations, *_lease_mgr, *_host_data_source_mgr, the implementation of the ipv4 is always before the ipv6. This was the only inconsistency.
It would be nice to order the functions in the cpp file to match the order of the definitions in the header, but this might be really to much for you. It could be done, but it would be harder to maintain and it would require some mentioning in the coding guidelines.

Thanks for the effort!
Razvan

tomaszmrugalski commented:
Thanks a lot for this detailed explanation. I suspected that you may have some extra code and that was the reason for odd wraps with status. That's why I merged the code as you proposed.

The reason why I was complaining about those changes is that we're doing some modifications related to RADIUS and I was afraid there may be conflicts. Both functionalities - Cassaandra and Radius - are complex, so it would be very tricky to check that the merge between them is done properly. So the best solution is to avoid having the problem altogether.

But I looked at the Radius changes in a bit more detail and they doesn't seem to cause conflicts after all. So it's good.
